echo_server.py

```diff
@@ -1,28 +1,149 @@
 #! /usr/bin/env python
 """Echo server in socket connection: receives and sends back a message."""
 import socket
+import os
+import mimetypes
+import binascii
+
+
+def response_ok(tuple):
+    """Return byte string 200 ok response."""
+    # u"HTTP/1.1 200 OK\r\nContent-Type: text/plain\r\nContent-length: 18\r\n\r\neverything is okay".encode('utf-8')
+    okay = u"HTTP/1.1 200 OK\r\nContent-Type: "
+    c_len = u"\r\nContent-length: "
+    crlf = u"\r\n\r\n"
+    return "{}{}{}{}{}{}".format(okay, tuple[0], c_len, len(tuple[1]), crlf, tuple[1])
+
+def response_error(error_code, reason):
+    """Return byte string error code."""
+    return u"HTTP/1.1 {} {}\r\nContent-Type: text/plain\r\nContent-length: {}\r\n\r\n{}".format(error_code, reason, len(reason), reason).encode('utf-8')
+
+
+# def parse_request(request):
+#     """Parse HTTP request and return URI requested.
+#     If not GET requests, raises error.
+#     If not HTTP/1.1 request, raises error."""
+#     first_line = request.split("\n")[0].split()
+#     if first_line[2] == 'HTTP/1.1':
+#         if first_line[0] == 'GET':
+#             return first_line[1]
+#         else:
+#             return response_error(405, 'Method Not Allowed')
+#     else:
+#         return response_error(505, 'HTTP Version Not Supported')
+
+class Error404(BaseException):
+    def __init__(self):
+        self.out = response_error(404, '404 Not Found')
+
+    def __str__(self):
+        return self.out
+
+
+class Error405(BaseException):
+    def __init__(self):
+        self.out = response_error(405, '405 Method Not Allowed')
+
+    def __str__(self):
+        return self.out
+
+
+class Error505(BaseException):
+    def __init__(self):
+        self.out = response_error(505, '505 HTTP Version Not Supported')
+
+    def __str__(self):
+        return self.out
+
+
+def parse_request2(request):
+    """Parse HTTP request and returns first line as a list."""
+    return request.split("\n")[0].split()
+
+
+def check_request(first_line):
+    """Check HTTP 1.1 and GET, then return uri."""
+    print "first_line: {}".format(first_line)
+    if first_line[2] != 'HTTP/1.1':
+        raise Error505
+    if first_line[0] != 'GET':
+        raise Error405
+    return first_line[1]
+
+
+def resource_uri(uri):
+    """Given URI, returns tuple with a type header and a body.
+    If directory, returns html listing of directory.
+    If file, returns content of file as body.
+    If resource not found, raise Error404.
+    """
+    # if asking for root change uri to '.'
+    if uri == '/':
+        uri = '.'
+    # remove / from beginning of string
+    if uri[0] == '/':
+        uri = uri[1:]
+    # Determine if directory, file or not found
+    if os.access(uri, os.F_OK):
```

Comment:
EAFP

it is discouraged using `if os.access(uri, os.F_OK)` as a test because it is possible that between asking and then executing the conditions could change.  Just try reading and if it raises an error, deal with it.